plugins/survey/src/types.ts

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright © 2023 Hardcore Engineering Inc.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import { AttachedDoc, Class, CollectionSize, Doc, Markup, Obj, Ref, Space } from '@hcengineering/core'
+import { ComponentType, SvelteComponent } from 'svelte'
+import type { Resource } from '@hcengineering/platform'
+
+/**
+ * @public
+ *
+ * @link https://www.npmjs.com/package/lexorank
+ */
+export type Rank = string
+
+/**
+ * @public
+ */
+export interface Survey extends Space {
+  questions: CollectionSize<Question>
+}
+
+/**
+ * @public
+ */
+export interface Question<Data extends QuestionData = QuestionData> extends AttachedDoc {
+  attachedTo: Ref<Survey>
+  attachedToClass: Ref<Class<Survey>>
+  rank: Rank
+  data: Data
+}
+
+/**
+ * @public
+ *
+ * Base marker interface for every question data implementation
+ */
+export interface QuestionData extends Obj {}
+
+/**
+ * @public
+ */
+export interface RadioButtonsOption {
+  label: string
+}
+
+/**
+ * @public
+ */
+export interface RadioButtons extends QuestionData {
+  text: string
+  options: RadioButtonsOption[]
+}
+
+/**
+ * @public
+ */
+export interface CheckboxesOption {
+  label: string
+}
+
+/**
+ * @public
+ */
+export interface Checkboxes extends QuestionData {
+  text: string
+  options: CheckboxesOption[]
+}
+
+/**
+ * @public
+ */
+export interface Info extends QuestionData {
+  text: Markup
+}
+
+/**
+ * @public
+ */
+export interface QuestionDataEditorComponentProps<Q extends QuestionData> {
+  readonly object: Q
+  readonly editable: boolean
+  readonly submit: (data: Q) => Promise<void>
+}
+
+/**
+ * @public
+ */
+export type QuestionDataEditorComponent<Q extends QuestionData> = SvelteComponent<QuestionDataEditorComponentProps<Q>>
+
+/**
+ * @public
+ */
+export type QuestionDataEditorComponentTypeRef<Q extends QuestionData> = Resource<
+ComponentType<QuestionDataEditorComponent<Q>>
+>
+
+/**
+ * @public
+ */
+export interface QuestionDataEditor<Q extends QuestionData = QuestionData> extends Class<Doc> {
+  editor: QuestionDataEditorComponentTypeRef<Q>
+}
```

Comment:
Some TS magic so that TS could help us control typings in question data editors – they must implement pretty specific and exact interface, and that's how we achieve this